include/swift/AST/Attr.h

```diff
@@ -1672,6 +1672,13 @@ class DifferentiableAttr final
   /// attribute's where clause requirements. This is set only if the attribute
   /// has a where clause.
   GenericSignature DerivativeGenericSignature;
+  /// The source location of the implicitly inherited protocol requirement
+  /// `@differentiable` attribute. Used for diagnostics, not serialized.
+  ///
+  /// This is set during conformance type-checking, only for implicit
+  /// `@differentiable` attributes created for non-public protocol witnesses of
+  /// protocol requirements with `@differentiable` attributes.
+  SourceLoc ImplicitlyInheritedDifferentiableAttrLocation;
```

Comment:
Note: it's possible to generalize this "diagnostic source location" for other kinds of implicit `@differentiable` attributes too, not just "ones that are implicitly inherited from protocol requirements".

Currently, there are only three kinds of implicit `@differentiable` attributes:
1. This case: `@differentiable` attributes inherited from protocol requirements.
2. `@differentiable` attributes synthesized from others on the same declaration with superset parameter indices.
    - `@differentiable(wrt: (x, y))` -> `@differentiable(wrt: x)`
3. `@differentiable` attributes synthesized on stored properties of structs/classes that derive a conformance to `Differentiable`.
```cpp
enum class ImplicitKind {
  InheritedFromProtocolRequirement, // this case
  SynthesizedSubsetParametersAttribute, // triggered during conformance type-checking
  SynthesizedForStoredProperty, // triggered during `Differentiable` derived conformances
};
```

I think the first case benefits the most from a "secondary diagnostic source location". Generalization could be done later.